osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs

```diff
@@ -77,11 +74,20 @@ protected override double StrainValueOf(DifficultyHitObject current)
             if (isOverlapping)
                 holdAddition = 1 / (1 + Math.Exp(0.5 * (release_threshold - closestEndTime)));
 
-            // Increase individual strain in own column
+            // Decay and increase individualStrains in own column
+            individualStrains[column] = applyDecay(individualStrains[column], startTime - startTimes[column], individual_decay_base);
             individualStrains[column] += 2.0 * holdFactor;
-            individualStrain = individualStrains[column];
 
-            overallStrain = applyDecay(overallStrain, current.DeltaTime, overall_decay_base) + (1 + holdAddition) * holdFactor;
+            // individualStrain should be the hardest individualStrain column for notes in a chord
+            individualStrain = maniaCurrent.DeltaTime == 0 ? Math.Max(individualStrain, individualStrains[column]) : individualStrains[column];
```

Comment:
Based on the comment, shouldn't this be the maximum over all columns? Like:

```
if (maniaCurrent.DeltaTime == 0)
    individualStrain = individualStrains.Max();
else
    individualStrain = individualStrains[column];
```

Otherwise, I'm not understanding the comment.

Reply:
As an aside, I'm not sure about checking against 0 since lazer uses floating point timing. I'd feel safer checking against `<= 1`.

Reply:
> Based on the comment, shouldn't this be the maximum over all columns?

We should only consider the notes that are actually being pressed here. The max part makes sure that new `individualStrains` aren't overriding old ones, which is what this bugfix is about. I updated the comment to make this more clear.

> I'm not sure about checking against 0 since lazer uses floating point timing. I'd feel safer checking against `<= 1`.

If the `deltaTime` isn't 0, then notes should be guaranteed to be ordered. But, if it makes you sleep at night, I guess it shouldn't have too much of an effect.

Reply:
> We should only consider the notes that are actually being pressed here

In that case I would still expect the following:
```csharp
// For notes at the same time (in a chord), the individualStrain should be the hardest individualStrain out of those columns
if (maniaCurrent.DeltaTime <= 1)
{
    for (int i = 0; i < startTimes.Length; i++)
    {
        if (Math.Abs(maniaCurrent.StartTime - startTimes[i]) > 1)
            continue;

        individualStrain = Math.Max(individualStrain, individualStrains[i]);
    }
}
else
    individualStrain = individualStrains[column];
```

Let me explain... The current code in this PR is just taking on the last strain for the _current_ column. You're not actually doing the "should be the hardest individualStrain out of those columns" part of what this comment says.

I believe it's fixing the issue by chance, but missing the point.

Furthermore, if the intent is as per my example above, the order in which these strains are processed becomes (or, still is) important. I don't think there is actually a way to handle this with how this skill is set up - it would need to be row-based.

Reply:
> You're not actually doing the "should be the hardest individualStrain out of those columns" part of what this comment says.

IndividualStrain is the last column's individualStrain that was processed. In the case where a note appears at the same time, it's comparing that last column to the current column, and taking the maximum. If there's more notes, it will just compare the previous maximum to the new current note. By the end, we have the hardest individualStrain in that chord. So, I'm not sure why you think that.

> I believe it's fixing the issue by chance, but missing the point.

I'm not totally sure why there's a difference between your code above and the code currently in this PR. Conceptually, it should be doing the same thing, but when testing it on my set of test beatmaps, it doesn't work the same way. Honestly quite lost myself.

For reference, here are my test beatmaps. Should be rather obvious which maps should be the same star rating (although my difficulty names aren't great).
[test beatmaps.zip](https://github.com/ppy/osu/files/9349395/test.beatmaps.zip)

Reply:
Right, I see. That makes sense.

> but when testing it on my set of test beatmaps, it doesn't work the same way

I think it's because it doesn't consider if `individualStrains[i]` is already `> individualStrain` for the first object. Basically it should only be maximum-ing the columns that have already been processed.

That being said, I do believe it's still order-dependent unless I'm missing something. Assume the individual strain of a 3 column chord `[ 1.0, 2.0, 3.0 ]`. If the update happens left-to-right, we get iS: `1.0 -> 2.0, 3.0`. If the update happens right-to-left, we get iS `3.0 -> 3.0 -> 3.0`.

Reply:
Maybe one way to fix this is to sort the columns?
```diff
diff --git a/osu.Game.Rulesets.Mania/Difficulty/ManiaDifficultyCalculator.cs b/osu.Game.Rulesets.Mania/Difficulty/ManiaDifficultyCalculator.cs
index 5d30d33190..f9bdf59afb 100644
--- a/osu.Game.Rulesets.Mania/Difficulty/ManiaDifficultyCalculator.cs
+++ b/osu.Game.Rulesets.Mania/Difficulty/ManiaDifficultyCalculator.cs
@@ -69,9 +69,10 @@ private static int maxComboForObject(HitObject hitObject)

         protected override IEnumerable<DifficultyHitObject> CreateDifficultyHitObjects(IBeatmap beatmap, double clockRate)
         {
-            var sortedObjects = beatmap.HitObjects.ToArray();
-
-            LegacySortHelper<HitObject>.Sort(sortedObjects, Comparer<HitObject>.Create((a, b) => (int)Math.Round(a.StartTime) - (int)Math.Round(b.StartTime)));
+            var sortedObjects = beatmap.HitObjects
+                                       .OrderBy(h => (int)h.StartTime)
+                                       .ThenBy(h => ((ManiaHitObject)h).Column)
+                                       .ToArray();

             List<DifficultyHitObject> objects = new List<DifficultyHitObject>();
```
The only reason it is using an unstable sort was to match osu!stable back when the diffcalcs were being reimplemented.

Reply:
> I do believe it's still order-dependent unless I'm missing something.

I believe that you're missing the fact that these notes occur at the same time, meaning they are part of the same strain section. Since strain only cares about the maximum, as long as the final note in the chord has the correct strain, it shouldn't matter what order we process them in.

> Maybe one way to fix this is to sort the columns?

I did consider this solution as well, but the problem here is that now the last column is going to take priority in difficulty calculation, meaning that reordering columns will still change star rating (either by mappers wanting to increase the star rating of their map or just mirroring it). It does solve the issue of note ordering if the notes are in the same places, but I feel this is not really tackling the core problem like my PR does. Also, if my PR really does solve the note ordering issue we shouldn't need to use the same unstable sort that stable does, if another sort happens to be faster.

Reply:
Right, I understand how this works now. It's because of this very tricky subtraction of `CurrentStrain` at the end of this method:
https://github.com/ppy/osu/blob/eb9cfe97d2aae2915f5bb084a52e726dc2bee58b/osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs#L86

So basically this also means that we don't need to sort objects anymore at all, correct? I really want to remove that sorting function because it's using probably the most hacky code in the entire repo.

Reply:
Sure, sounds good to me. I'll let you make that change since I don't really want to touch anything outside of the main difficulty calculation.